src/java/org/apache/cassandra/index/sai/utils/TypeUtil.java

```diff
@@ -0,0 +1,482 @@
+import org.apache.cassandra.db.marshal.AsciiType;
+import org.apache.cassandra.db.marshal.BooleanType;
+import org.apache.cassandra.db.marshal.ByteBufferAccessor;
+import org.apache.cassandra.db.marshal.CollectionType;
+import org.apache.cassandra.db.marshal.CompositeType;
+import org.apache.cassandra.db.marshal.DecimalType;
+import org.apache.cassandra.db.marshal.InetAddressType;
+import org.apache.cassandra.db.marshal.IntegerType;
+import org.apache.cassandra.db.marshal.ReversedType;
+import org.apache.cassandra.db.marshal.UTF8Type;
+import org.apache.cassandra.db.rows.Cell;
+import org.apache.cassandra.db.rows.ComplexColumnData;
+import org.apache.cassandra.index.sai.plan.Expression;
+import org.apache.cassandra.schema.ColumnMetadata;
+import org.apache.cassandra.serializers.MarshalException;
+import org.apache.cassandra.utils.ByteBufferUtil;
+import org.apache.cassandra.utils.FastByteOperations;
+import org.apache.cassandra.utils.bytecomparable.ByteComparable;
+import org.apache.cassandra.utils.bytecomparable.ByteSource;
+import org.apache.cassandra.utils.bytecomparable.ByteSourceInverse;
+
+public class TypeUtil
+{
+    private static final byte[] IPV4_PREFIX = new byte[] { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, -1, -1 };
+
+    /**
+     * DecimalType / BigDecimal values are indexed by truncating their asComparableBytes representation to this size,
+     * padding on the right with zero-value-bytes until this size is reached (if necessary).  This causes
+     * false-positives that must be filtered in a separate step after hitting the index and reading the associated
+     * (full) values.
+     */
+    public static final int DECIMAL_APPROXIMATION_BYTES = 24;
+
+    private TypeUtil() {}
+
+    /**
+     * Returns <code>true</code> if given buffer would pass the {@link AbstractType#validate(ByteBuffer)}
+     * check. False otherwise.
+     */
+    public static boolean isValid(ByteBuffer term, AbstractType<?> validator)
+    {
+        try
+        {
+            validator.validate(term);
+            return true;
+        }
+        catch (MarshalException e)
+        {
+            return false;
+        }
+    }
+
+    /**
+     * Indicates if the type encoding supports rounding of the raw value.
+     *
+     * This is significant in range searches where we have to make all range
+     * queries inclusive when searching the indexes in order to avoid excluding
+     * rounded values. Excluded values are removed by post-filtering.
+     */
+    public static boolean supportsRounding(AbstractType<?> type)
+    {
+        return isBigInteger(type) || isBigDecimal(type);
+    }
+
+    /**
+     * Returns the smaller of two {@code ByteBuffer} values, based on the result of {@link
+     * #compare(ByteBuffer, ByteBuffer, AbstractType)} comparision.
+     */
+    public static ByteBuffer min(ByteBuffer a, ByteBuffer b, AbstractType<?> type)
+    {
+        return a == null ?  b : (b == null || compare(b, a, type) > 0) ? a : b;
```

Comment:
Nit: extra whitespace in the ternary operator of all the `min`/`max` versions:
```suggestion
        return a == null ? b : (b == null || compare(b, a, type) > 0) ? a : b;
```